File: 01/main.c
#include <assert.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>


typedef struct {
    int first_num;
    int last_num;
} NumberPair;

char * get_input(const char *filepath);
bool is_digit(char c);
int part1(char *input);
int part2(char *input);
int is_word_digit(char *input);
int get_second_number_for_part2(char *input);
NumberPair get_first_and_last_number_with_word(char * line, int line_len);
// int combine_numbers(int first_num, int second_num);

#define MAX_FILE_SIZE (24 * 1024)
char input[MAX_FILE_SIZE] = {0};


typedef struct {
    char *number;
    int len;
} Number;

#define NUMBERS_LEN 9
Number NUMBERS[NUMBERS_LEN] = 
{
    { .number = "one", .len = 3, },
    { .number = "two", .len = 3, },
    { .number = "three", .len = 5, },
    { .number = "four", .len = 4, },
    { .number = "five", .len = 4, },
    { .number = "six", .len = 3, },
    { .number = "seven", .len = 5, },
    { .number = "eight", .len = 5, },
    { .number = "nine", .len = 4}
};

int main(void) {
    const char *filepath = "./input.txt";
    char *input = get_input(filepath);
    if (input == NULL)
    {
        return 1;
    }
    int sum1 = part1(input);
    printf("part1: total sum: %i\n", sum1);
    int sum2 = part2(input);
    printf("part2: total sum: %i\n", sum2);
    return 0;
}

int part2(char *input) {
    int sum = 0;
    int line_count = 0;
    for (int i = 0; input[i] != '\0'; ++i) {
        char line[256] = {0};
        int line_len = 0;
        for (; input[i] != '\0' && input[i] != '\n'; ++line_len, ++i) {
            line[line_len] = input[i];
        }
        NumberPair np = get_first_and_last_number_with_word(line, line_len);

        int two_digit_num = np.first_num * 10 + np.last_num;
        sum += two_digit_num;
        line_count++;
    }
    return sum;
}

int part1(char *input) {
    int sum = 0;
    int first_num = 0;
    int second_num = 0;
    for (int i = 0; input[i] != '\0'; ++i) {
        if (first_num == 0 && is_digit(input[i])) {
            first_num = input[i] - '0';
            second_num = first_num;
            continue;
        }
        if (is_digit(input[i])) {
            second_num = input[i] - '0';
        }
        if (input[i] == '\n') {
            int two_digit_num = first_num * 10 + second_num;
            sum += two_digit_num;
            first_num = 0;
            second_num = 0;
        }
    }
    return sum;
}

char * get_input(const char *filepath) {
    FILE *f = fopen(filepath, "r");
    if (f == NULL) {
        perror("ERROR");
        return NULL;
    }
    int ret = fseek(f, 0, SEEK_END);
    assert(ret == 0 && "fseek failed");
    size_t filesize = ftell(f);
    assert(filesize <= MAX_FILE_SIZE);
    rewind(f);
    ret = fread(input, filesize, 1, f);
    // assert(ret == 1);
    if (!feof(f) && ferror(f)) {
        perror("ERROR: ");
        return NULL;
    }
    fclose(f);
    return input;
}

bool is_digit(char c) {
    return c >= '0' && c <= '9';
}

int is_word_digit(char *input) {
    for (int i = 0; i < NUMBERS_LEN; ++i) {
        if (strncmp(NUMBERS[i].number, input, NUMBERS[i].len) == 0) {
            // printf("number: %s, input: %s\n", NUMBERS[i].number, input);
            return i + 1;
        }
    }
    return 0;
}

int get_second_number_for_part2(char *input) {
    char line[256] = {0};
    int line_len = 0;
    for (int i = 0; *input != '\0' && *input != '\n'; ++i) {
        line[i] = *input++;
        line_len++;
    }
    while (line_len > 3) {
        char last_char = line[line_len - 1];
        if (is_digit(last_char)) {
            return last_char - '0';
        }
        for (int i = 0; i < NUMBERS_LEN; ++i) {
            if (line_len < NUMBERS[i].len) {
                continue;
            }
            char *word_from_last = (char *) line + line_len - NUMBERS[i].len;
            if (strncmp(word_from_last, NUMBERS[i].number, NUMBERS[i].len) == 0) {
                return i + 1;
            }
        }
        line_len--;
    }
    return 0;
}

NumberPair get_first_and_last_number_with_word(char * line, int line_len) {
    int first_num = 0;
    int second_num = 0;
    for (int i = 0; i < line_len; ++i) {
        char c = line[i];
        if (is_digit(c)) {
            if (first_num == 0) {
                first_num = c - '0';
                second_num = first_num;
            } else {
                second_num = c - '0';
            }
        } else {
            char * word = ((char *) line) + i;
            for (int k = 0; k < NUMBERS_LEN; ++k) {
                Number num = NUMBERS[k];
                if (line_len - i < num.len) {
                    continue;
                }

                if (strncmp(num.number, word, num.len) == 0) {
                    if (first_num == 0) {
                        first_num = k + 1;
                        second_num = first_num;
                    } else {
                        second_num = k + 1;
                    }
                    break;
                }
            }
        }
    }

    return (NumberPair) {
        .first_num = first_num,
        .last_num = second_num,
    };
}

File: 01/main.rs
use std::fs;

struct Number (String, usize);


fn main() {
    let filepath = "./input.txt";
    let input = fs::read_to_string(filepath).unwrap();

    let numbers: Vec<Number> = vec![
        Number(String::from("one"),   3,),
        Number(String::from("two"),   3,),
        Number(String::from("three"), 5,),
        Number(String::from("four"),  4,),
        Number(String::from("five"),  4,),
        Number(String::from("six"),   3,),
        Number(String::from("seven"), 5,),
        Number(String::from("eight"), 5,),
        Number(String::from("nine"),  4,),
    ];
    part1(input.clone());
    part2(input.clone(), &numbers);
}

#[allow(dead_code)]
fn part1(input: String) {
    let mut total_sum: usize = 0;

    for line in input.lines() {
        let (first_num, second_num) = get_first_and_last(line);
        let combined = first_num * 10 + second_num;
        total_sum += combined;
    }
    println!("part1: total_sum: {}", total_sum);
}

fn part2(input: String, numbers: &Vec<Number>) {
    let mut total_sum: usize = 0;

    for line in input.lines() {
        let (first_num, second_num) = get_first_and_last_with_word_support(line, numbers);
        let combined = first_num * 10 + second_num;
        total_sum += combined;
    }
    println!("part1: total_sum: {}", total_sum);
}

fn get_first_and_last_with_word_support(line: &str, numbers: &Vec<Number>) -> (usize, usize) {
    let mut first_num: usize = 0;
    let mut second_num: usize = 0;
    let line_len = line.len();
    for (i, ch) in line.chars().enumerate() {
        if is_digit(ch) {
            if first_num == 0 {
                first_num = ch as usize - '0' as usize;
                second_num = first_num;
            } else {
                second_num = ch as usize - '0' as usize;
            }
        } else {
            for (k, number) in numbers.iter().enumerate() {
                let end = if i + number.1 > line_len { line_len } else { i + number.1 };
                let slice = &line[i .. end];
                if slice == number.0 {
                    if first_num == 0 {
                        first_num = k + 1;
                        second_num = first_num;
                    } else {
                        second_num = k + 1;
                    }
                    break;
                }
            }
        }
    }
    return (first_num, second_num);
}

fn get_first_and_last(line: &str) -> (usize, usize) {
    let mut first_num: usize = 0;
    let mut second_num: usize = 0;
    for ch in line.chars() {
        if is_digit(ch) {
            if first_num == 0 {
                first_num = ch as usize - '0' as usize;
                second_num = ch as usize - '0' as usize;
            } else {
                second_num = ch as usize - '0' as usize;
            }
        }
    }
    return (first_num, second_num);
}


fn is_digit(ch: char) -> bool {
    return ch >= '0' && ch <= '9';
}

#[allow(dead_code)]
fn print_type_of<T>(_: &T) {
    println!("{0}", std::any::type_name::<T>());
}
